File: LC/Medium/2061.py
class Solution:
    def numberOfCleanRooms(self, room: List[List[int]]) -> int:
        rows = len(room)
        cols = len(room[0])
        curr_dir = 0
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        r = 0
        c = 0
        visited = set()
        room[r][c] = -1
        ans = 1
        
        while (r, c, curr_dir) not in visited:
            visited.add((r, c, curr_dir))
            n_r = r + dirs[curr_dir][0]
            n_c = c + dirs[curr_dir][1]
            
            if 0 <= n_r < rows and 0 <= n_c < cols and room[n_r][n_c] != 1:
                r, c = n_r, n_c
                if room[r][c] == 0:
                    ans += 1
                    room[r][c] = -1
            else:
                curr_dir = (curr_dir + 1) % 4
        return ans

File: LC/C++/Medium/2061.cpp
class Solution {
    std::vector<std::vector<int>> dirs{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
public:
    int numberOfCleanRooms(vector<vector<int>>& room) {
        int m = room.size(), n = room[0].size();
        int dirr = 0; 
        int r = 0;
        int c = 0;
        vector<vector<vector<bool>>> visited(
            m, vector<vector<bool>>(n, vector<bool>(4, false))
        );
        room[r][c] = -1;
        int ans = 1;
        while (!visited[r][c][dirr]) {
            visited[r][c][dirr] = true;
            int nr = r + dirs[dirr][0];
            int nc = c + dirs[dirr][1];
            
            if (0 <= nr && nr < m && 0 <= nc && nc < n && room[nr][nc] != 1) {
                r = nr;
                c = nc;
                if (room[r][c] == 0) {
                    room[r][c] = -1;
                    ans++;
                }
            } else {
                dirr = (dirr + 1) % 4;
            }
        }

        return ans;
    }
};